### PR TITLE
Update signup progress state to a keyed object

### LINKED ISSUE #32506: Signup: broken transition from business to ecommerce flow
**Steps to reproduce:**
1. Go to `/start/business`
2. Enter "Sell" in the site goal box in the `about` step
3. Submit the step

You are redirected to `/start/ecommerce/domains` and that renders an empty page.

**Some analysis what happens:**
The `<Signup>` component [renders nothing](https://github.com/Automattic/wp-calypso/blob/9a85029645bd8b53ff313f1f3c77ad7281608dce/client/signup/main.jsx#L593-L599) when it thinks that a "flow resume" (abandoning the signup and coming back to the start later, we proceed to the first incomplete step) is in progress. The condition is:
```js
if ( this.getPositionInFlow() > 0 && this.props.progress.length === 0 ) {
  return null;
}
```

In our case, the position in flow is indeed >0, and the progress is empty for some reason.

The progress is empty because the `removeUnneededSteps` reducer is called when changing flow name, and it's very picky about the steps it keeps: the steps must have the same name, but also the same index in the array!

The progress array from the `business` flow contains the data for the `about` step, but it's not at the expected index. There is `plans-business` step before it, which was auto-fulfilled by the `removeFulfilledSteps` code.

Therefore, the filtered progress misses the `about` step (which is probably unwanted) and is empty. Then it triggers the "render nothing" condition.

Cc @niranjan-uma-shankar as he's the most recent committer of almost all the involved code.

### PR DESCRIPTION
**Pull request by spen**
> ~Note: This should be landed after #34770 is landed to avoid conflicts~

#### Changes proposed in this Pull Request

* Convert the signup progress store from an ordered array to an object, keyed by the names of the steps progressed.
* Access the relevant progress state according to the current flow.

#### Some notes:

Rather than the progress state being an exact representation of the steps currently progressed in the current flow, this approach attempts to make the progress store a collection of any steps that have been progressed during the signup journey.
Currently, when we switch flow we have to run logic on the progress store in an attempt to remove steps that are no longer needed, which is hard to do accurately and is causing bugs. It also means that we'd eagerly get rid of a step's progress if it didn't apply to the incoming flow, though it might become relevant again if the customer were to change their mind and switch back to that original flow. In this new approach we would hold on to the progress of every step that the customer has seen, regardless of its present usefulness and wouldn't need to attempt to keep the progress in sync with the flows.
In order to display the `3 out of 6 steps` information we just use the list of step names provided by the flow definition and match up the relevant steps found in the progress store. 

There are a few more notes over at #34885 (superseded by this PR) 

#### Testing instructions

**Testing user step last A/B test behaviour**
- Please follow the test instructions found in #34770 - whether that PR is landed yet the behaviour should remain the same as described there.
- More test instructions to come soon...

Fixes #32506

### CODE REVIEW

**Comment by adamziel** [comment]
Just noting there are some conflicts and failing tests

**Comment by spen** [comment]
Thanks Adam, I'm hoping to get some 👀 on this approach before potentially wasting time fixing/writing tests if the team don't want to move in this direction, or if there's something fundamentally wrong with this approach 👍
Same goes for running on the rebase treadmill, I don't want to get in to that loop until we're confident we want to merge.

**Comment by p-jackson** [comment]
@spen fwiw I much prefer this approach (granted I've mainly looked at #35139). I'm not sure about calling it `progress` anymore though. That word does seem like a linear thing, so it makes me think it would be an array.

Would `stepData` or `stepState` be more accurate?

Also given so many components still expect the data in array form (and so there's many calls to `getFilteredSteps()`) maybe there should be a new selector for transforming the data into the shape expected by components? That way it's memoizable etc. etc., all the standard Redux tricks.

**Comment by spen** [comment]
> I'm not sure about calling it progress anymore though... Would stepData or stepState be more accurate?

Totally! I've been having similar thoughts too! To me, simply `signup.steps` would be ideal.
I think that at some point we should take a more holistic look at the state and figure out what should go where, what the correct semantics and what we can get rid of to simplify our state and make it more reliable.

> maybe there should be a new selector for transforming the data into the shape expected by components? That way it's memoizable etc. etc., all the standard Redux tricks.

Yeah! I'm surprised this didn't cross my mind to be honest, it felt very icky making so many calls to the filtering functions in components.

Now that I know this is a direction we're considering I'll put some extra work in to neaten this up! 👌

**Comment by p-jackson** [comment]
> To me, simply `signup.steps` would be ideal

Yeah definitely

File: client/state/test/initial-state.js
Comment (adamziel) [inline client/state/test/initial-state.js]: It would be nice to avoid duplication here - the key is the same as stepName. Let's not delay this PR though - just keep not of that for the future.

Comment (spen) [reply client/state/test/initial-state.js]: Totally! I'll look in to what it would take to do that soon after :)

**Review by adamziel — DISMISSED** [review]
Did you test this PR @spen ? The back button alternates between the previous and the current step:

https://cld.wthms.co/DTIqg8

~~Edit: This seems to be on me - I'm experiencing some quirks in my local calypso setup.~~
Edit 2: That's actually a problem with this PR

File: client/signup/utils.js
Comment (adamziel) [inline client/signup/utils.js]: What would happen if we started our journey in flow A but now we're in flow B?

**Comment by adamziel** [comment]
I pushed a fix for the back button problem, will keep on testing

File: client/state/signup/progress/reducer.js
Comment (adamziel) [inline client/state/signup/progress/reducer.js]: Nitpick: Using lodash for simple stuff like that adds an extra step of "better look it up in lodash docs" while reviewing. `step.stepName in state` accomplishes the same thing and is easier to read

File: client/state/signup/progress/reducer.js
Comment (adamziel) [inline client/state/signup/progress/reducer.js]: Same lodash nitpick as in other comment: `state[stepName]` does the same and is simpler. I've seen some more examples but you get the idea.

File: client/state/signup/progress/reducer.js
Comment (adamziel) [inline client/state/signup/progress/reducer.js]: So glad to see this gone

File: client/state/signup/progress/reducer.js
Comment (adamziel) [inline client/state/signup/progress/reducer.js]: This PR removes the reducer, but the constant and the action itself are still used in a few places. There are no other reducers handling this so it should be safe to remove any other references as well.

Comment (adamziel) [reply client/state/signup/progress/reducer.js]: By the way it's not a blocker for merging - it can totally be adressed in a subsequent PR

File: client/signup/navigation-link/test/index.jsx
Comment (adamziel) [inline client/signup/navigation-link/test/index.jsx]: getFilteredSteps returns an array, not an object - thus this tests for the wrong type of value

Comment (adamziel) [reply client/signup/navigation-link/test/index.jsx]: Also it would be great to have some tests for getFilteredSteps itself - this PR seems to revolve around it, but it has no coverage

**Review by adamziel — DISMISSED** [review]
I left some notes, but other than fixing tests I don't see any major blockers. I'd love to get more eyes on this PR and if another person says it works and doesn't introduce any problems - I think it's good to go!

As a side note, I think this even fixes this bug: https://github.com/Automattic/wp-calypso/issues/35363

File: client/signup/navigation-link/index.jsx
Comment (p-jackson) [reply client/signup/navigation-link/index.jsx]: nbd, just a thought

**Review by p-jackson — COMMENTED** [review]
I agree with Adam's comments too, but in general this looks good and I far prefer this store layout. Proceeding to some manual testing.

**Review by p-jackson — COMMENTED** [review]
I agree with Adam's comments too, but in general this looks good and I far prefer this store layout. Proceeding to some manual testing.

File: client/signup/navigation-link/index.jsx
Comment (adamziel) [reply client/signup/navigation-link/index.jsx]: @p-jackson ah yes, that was me not adjusting line 59 after introducing line 58 - good spot!

**Comment by p-jackson** [comment]
Tests well for me. Also went through the cases in https://github.com/Automattic/wp-calypso/pull/34211#issuecomment-506298663 👍

File: client/state/signup/progress/reducer.js
Comment (spen) [reply client/state/signup/progress/reducer.js]: Thanks for spotting this Adam! I've opened a follow up PR here: #35461
Once I finish up a test plan and once this is merged I'll publish as a proper PR :)

File: client/signup/navigation-link/test/index.jsx
Comment (spen) [reply client/signup/navigation-link/test/index.jsx]: Good point, I'll add some today, thanks @adamziel !

File: client/blocks/login/index.jsx
Comment (spen) [reply client/blocks/login/index.jsx]: Good call and I do actually agree, I'll fix this up :)

File: client/signup/utils.js
Comment (spen) [reply client/signup/utils.js]: Ahh well, if they have completed 3 steps from flow A, then switched to flow B and flow B shared 2 out the 3 steps from flow A, those 2 steps would be held in state (as is the 1 unshared step, actually) so when we call `getFilteredSteps` after switching to flow B 2 of it's steps will returned .

File: client/signup/navigation-link/test/index.jsx
Comment (spen) [reply client/signup/navigation-link/test/index.jsx]: Thanks for spotting that! - `getFilteredSteps` should now be pretty well covered in https://github.com/Automattic/wp-calypso/pull/35049/commits/def07a343be80fb790ad6c773eb35c5bd991346b

File: client/signup/navigation-link/index.jsx
Comment (adamziel) [inline client/signup/navigation-link/index.jsx]: Why would that not return an array though? It seems like this check would conceal a legitimate bug if it were to happen one day

File: client/signup/navigation-link/index.jsx
Comment (adamziel) [inline client/signup/navigation-link/index.jsx]: As @p-jackson mentioned, we could remove `step.stepName !== stepName` as that's accounted for in the `slice` in the previous line

File: client/state/signup/progress/reducer.js
Comment (spen) [reply client/state/signup/progress/reducer.js]: I'll leave this one be, this suggestion would error in the case where `state` is `undefined`.
I know thats hugely unlikely, but I always prefer to be consistent with patterns like this.
Also, I'd argue that any dev has to look up documentation when touching new code and we lodash is well used throughout Calypso, so the learning opportunity wouldn't hurt and could encourage others to use safer methods.

File: client/blocks/login/index.jsx
Comment (spen) [reply client/blocks/login/index.jsx]: Done in https://github.com/Automattic/wp-calypso/pull/35049/commits/00edebe1d64019fb48d2c1d0f4781427a5d2e550 :)

File: client/signup/navigation-link/index.jsx
Comment (spen) [reply client/signup/navigation-link/index.jsx]: Taken care of in https://github.com/Automattic/wp-calypso/pull/35049/commits/b569dab205e97c98d816ee251ab1c5b9bd697620

File: client/signup/navigation-link/index.jsx
Comment (spen) [reply client/signup/navigation-link/index.jsx]: Yeah, we could have ended up returning something else from `getFilteredSteps` and have been returning `{ stepName: null }` constantly which would have been tricky to debug - good catch!
I've removed this check 👍

File: client/signup/test/utils.js
Comment (p-jackson) [inline client/signup/test/utils.js]: Personally prefer writing these sort of object/array tests using `toEqual` which does a deep equality check.
```javascript
expect( result ).toEqual( [ userStep, siteTypeStep ] );
```

Using `toContain` has better semantics, which I like, but using `toEqual` will also test:
- elements are in the correct order
- there aren't other random elements in there
- doubles as an `isArray` test too :)

**Review by p-jackson — COMMENTED** [review]
Changes look good 👍
I pushed a wee fix to the `<NavigationLink>` unit tests; mocking meant they weren't running independently.

Failing canary test is signup related so I'll take a peek to see how real the issue is.

**Comment by p-jackson** [comment]
Yup, e2e failures are real. Simplified steps for reproducing bug:

1. Start flow that pre-selects a plan (e.g. `/start/personal`)
2. Go through steps, any site type, any domain
3. Gets stuck on the "Give us one minute and we’ll move right along" animation forever.

**Comment by ramonjd** [comment]
> Yup, e2e failures are real. 

Could it have anything to do with the pre-selected plan issue I mentioned in #35139 
?


https://github.com/Automattic/wp-calypso/pull/35139#discussion_r311359757

File: client/lib/signup/flow-controller.js
Comment (andrewserong) [inline client/lib/signup/flow-controller.js]: The A/B test to move the user step has wrapped up (https://github.com/Automattic/wp-calypso/pull/35552), so this line can be removed now.

File: client/lib/signup/test/flow-controller.js
Comment (andrewserong) [inline client/lib/signup/test/flow-controller.js]: This is just a tiny style nit, but It looks like a common pattern in other tests in the codebase for checking for an empty object is to use Jest's `.toEqual( {} );` so the lodash `isEmpty` might be unnecessary here and on line 86.

File: client/state/signup/progress/reducer.js
Comment (andrewserong) [inline client/state/signup/progress/reducer.js]: Feel free to disregard this one, but it looks like you've refactored all of the other regular functions in this reducer to use arrow functions assigned to a `const`, but this one's still a regular function, so it stood out to me.

**Comment by p-jackson** [comment]
Yup, @ramonjd's hint in https://github.com/Automattic/wp-calypso/pull/35139#discussion_r311359757 was just what I needed. Pending excluded steps now working again.

I'll try rebasing this to remove the reference to the `moveUserStepPosition` @andrewserong pointed out.

**Comment by p-jackson** [comment]
I think this PR is all working now. I haven't addressed any of the minor points from the reviews, I'll let you take a look @spen before we 🤞 move to the next step!

File: client/state/signup/progress/reducer.js
Comment (spen) [reply client/state/signup/progress/reducer.js]: Thanks @andrewserong, I'll leave this one be for now though to avoid growing the diff.

**Comment by spen** [comment]
Awesome, thanks for sorting this @p-jackson ! I was stuck on the same issue yesterday too - I'd gone down a different route that wasn't so fruitful - the changes in https://github.com/Automattic/wp-calypso/pull/35049/commits/943ffdc89fd993180981d169e71745f9173b6a2a do a great job of solving this issue 👌

**Comment by adamziel** [comment]
@andrewserong @ramonjd would you mind giving this PR a spin and merging if it's all good now? It's an important piece of code as it will help Samus and Luna build Signup flows and AB tests with confidence.

**Comment by ramonjd** [comment]
Any particular flows we should test? I was going to run through site creation, jetpack connection... the usual suspects. 

Now that https://github.com/Automattic/wp-calypso/pull/34770 has been reverted and closed, we can safely ignore this PR's test description, right?

I'm still seeing the `moveUserStepPosition` AB test in this PR... going to rebase. 👍

**Comment by p-jackson** [comment]
@ramonjd no specific flows, what you mentioned sounds good. Things that interact with dependencies in weird and wonderful ways are probably the highest risks e.g. adding query parameters to flows, anything that auto-submits a step.

**Comment by p-jackson** [comment]
And yes these changes don't really have anything to do with #34770 anymore, so I'd disregard those parts.

**Comment by ramonjd** [comment]
Been through the following flows (successfully) so far....

- [ ] Login and create a site with onboarding flow
- [ ] Navigate between steps, changing between Business and Ecommerce site types, create an E-commerce step
- [ ] Go to /business and /premium flows, navigate between between Business and Ecommerce site types, select Professional and continue onto purchase
- [ ] art.blog: Create a site by clicking on CTA /start/user?site_type=blog&vertical=art&ref=dotblog-landing
- [ ] Try other permutations, e.g., /start/user?site_type=business&vertical=travel

**Review by andrewserong — APPROVED** [review]
Great work @spen — I've tested a number of different permutations of creating a site today, and haven't run into any issues. Looks like the failing e2e tests right now are Gutenberg related and not this PR, so personally I think this looks good to 🚢

I noticed that the bug https://github.com/Automattic/wp-calypso/issues/35363 is still present in this PR, but that isn't at all a reason to hold off on merging, and we can pursue a fix for that separately.

**Comment by ramonjd** [comment]
Also checked Jetpack connection flow. State checks out:

Screen Shot 2019-08-21 at 7 59 05 pm

Good work!

**Pull request merged**
